src/components/Diagram/Diagram.tsx

```diff
@@ -149,10 +151,28 @@ function toNodesAndEdges(definition: FlowDefinition | undefined) {
 
 export default function Diagram({ definition }: { definition: any }) {
   const nodeTypes = useMemo(() => ({ StateNode: StateNode }), []);
-  const { nodes, edges } = toNodesAndEdges(definition);
+  const { nodes: initialNodes, edges: initialEdges } = useMemo(
+    () => toNodesAndEdges(definition),
+    [definition],
+  );
+  const [nodes, setNodes, onNodesChange] = useNodesState(initialNodes);
+  const [edges, setEdges, onEdgesChange] = useEdgesState(initialEdges);
+
+  useEffect(() => {
+    setNodes(initialNodes);
+    setEdges(initialEdges);
+  }, [initialEdges, initialNodes, setEdges, setNodes]);
+
```

Comment:
Sorry, I might have led you a bit astray here based on how `definition` is used in this component; Since you are relying on the `definition` prop in `useMemo`, the combination of `useMemo` and `useEffect` is a bit superfluous (and actually might causing additional unnecessary renders) since both will run with the value changes.

I believe you can get the same desired behavior with just the `useEffect`:

```suggestion
  const [nodes, setNodes, onNodesChange] = useNodesState([]);
  const [edges, setEdges, onEdgesChange] = useEdgesState([]);

  useEffect(() => {
    const { nodes, edges } = toNodesAndEdges(definition);
    setNodes(nodes);
    setEdges(edges);
  }, [definition, setEdges, setNodes]);

```

That said, in both cases, I'm not sure if we should be resetting the layout when the definition changes. It doesn't feel like a super great UX when I move things around to my liking and then they are immediately reverted on any change in the editor: https://www.loom.com/share/703fae8ed4cb49cc8701d760d818b5bf?sid=34e878e2-8642-41f5-a202-42b67c876bcc

I'm trying to think of a decent approach here, but I keep coming back to we need to give users a way to toggle between the auto-layout behavior and the user-altered layout – lets catch up in the AM and see if we can figure something out!

Reply:
`  useEffect(() => {
    const { nodes: updatedNodes, edges: updatedEdges } = toNodesAndEdges(definition);
    updatedNodes.forEach(updatedNode => {
      const match = nodes.find(existingNode => existingNode.id === updatedNode.id);
      if (match) {
        // Retain position data associated with this node id
        updatedNode.position = match.position;
      }
    });
    setNodes(updatedNodes);
    setEdges(updatedEdges);
  }, [definition, setEdges, setNodes]);`

This is rough, but we could retain existing position data by node id to prevent that snap-back behavior when editing. We could also offer a a reset option to get the graph back to a default state if desired... just let me know your thoughts!